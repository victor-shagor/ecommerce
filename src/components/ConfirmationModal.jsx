import React from 'react';
import Button from 'react-bootstrap/Button';
import Modal from 'react-bootstrap/Modal';

const ConfirmationModal = ({text, handleConfirm, show, handleClose}) => {
 const handleYes = () =>{
  handleConfirm()
  handleClose()
 }
 
  return (
    <>
      <Modal show={show} onHide={handleClose} data-testid="confirmation-modal">
        <Modal.Header closeButton>
          <Modal.Title>Remove from cart</Modal.Title>
        </Modal.Header>
        <Modal.Body data-testid="modal-body">{text}</Modal.Body>
        <Modal.Footer>
          <Button variant="secondary" onClick={handleClose} data-testid="handle-close">
            Close
          </Button>
          <Button variant="primary" onClick={handleYes}>
            Yes
          </Button>
        </Modal.Footer>
      </Modal>
    </>
  );
}

export default ConfirmationModal