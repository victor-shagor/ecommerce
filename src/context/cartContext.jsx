import { createContext, useMemo, useState } from "react";
import {
  decreaseCartItemQuantity,
  increaseCartItemQuantity,
  removeCartItem,
} from "../utils/products";

export const CartContext = createContext(null);

const cartItemsFromStorage = JSON.parse(localStorage.getItem("cartItems"));

const setCartItemsToStorage = (cartItems) => {
  localStorage.setItem("cartItems", JSON.stringify(cartItems));
};

export const CartContextProvider = (props) => {
  const [cartItems, setCartItems] = useState(cartItemsFromStorage || []);
  const [showCart, setShowCart] = useState(false);

  const toggleCart = () => {
    setShowCart(!showCart);
  };

  const handleAddToCart = (product) => {
    const productExist = cartItems.find((el) => el.id === product.id);
    if (productExist) {
      const updatedCartProduct = increaseCartItemQuantity(product, cartItems);
      setCartItems(updatedCartProduct);
      setCartItemsToStorage(updatedCartProduct);
    } else {
      setCartItems([{ ...product, quantity: 1 }, ...cartItems]);
      setCartItemsToStorage([{ ...product, quantity: 1 }, ...cartItems]);
    }
    setShowCart(true);
  };

  const handleIncreaseQuantity = (product) => {
    const updatedCartProduct = increaseCartItemQuantity(product, cartItems);
    setCartItems(updatedCartProduct);
    setCartItemsToStorage(updatedCartProduct);
  };

  const handleDecreaseQuantity = (product) => {
    const updatedCartProduct = decreaseCartItemQuantity(product, cartItems);
    setCartItems(updatedCartProduct);
    setCartItemsToStorage(updatedCartProduct);
  };

  const handleRemoveCartItem = (product) => {
    const updatedCartProduct = removeCartItem(product, cartItems);
    setCartItems(updatedCartProduct);
    setCartItemsToStorage(updatedCartProduct);
  };

  const handleClearCart = () => {
    setCartItems([]);
    localStorage.removeItem("cartItems");
  };

  const cartPriceTotal = useMemo(() =>{
    const total = cartItems.reduce((acc, el)=> acc+(el.quantity*el.price),0).toFixed(2)
    let USDollar = new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
  });
    return USDollar.format(total)
  },[cartItems] )

  const contextValue = {
    cartItems,
    handleAddToCart,
    handleIncreaseQuantity,
    handleDecreaseQuantity,
    handleRemoveCartItem,
    toggleCart,
    showCart,
    handleClearCart,
    cartPriceTotal
  };

  return (
    <CartContext.Provider value={contextValue}>
      {props.children}
    </CartContext.Provider>
  );
};
