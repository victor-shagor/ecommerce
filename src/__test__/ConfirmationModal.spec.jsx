import {  fireEvent, render, screen } from "@testing-library/react";
import {  CartContextProvider } from "../context/cartContext";
import ConfirmationModal from "../components/ConfirmationModal";

jest.mock("axios", () => {
  return {
    __esModule: true
  }});

describe("Modal component", () => {

  it("should render confirmation modal", async () => {
    const handleClose = jest.fn()
    render(
      <CartContextProvider>
        <ConfirmationModal text='test body' show={true}  handleClose={handleClose} />
      </CartContextProvider>
    );

    expect(screen.getByTestId("confirmation-modal")).toBeInTheDocument();
    expect(screen.getByTestId("modal-body").innerHTML).toBe('test body');
  });

  it("confirmation modal should not render if show is false", async () => {
     render(
        <CartContextProvider>
        <ConfirmationModal show={false} />
      </CartContextProvider>
    );

    expect(()=>screen.getByTestId("confirmation-modal")).toThrow();
  });

  it(" handleclose should be called", async () => {
    const handleClose = jest.fn()
    const { findByTestId } = render(
        <CartContextProvider>
        <ConfirmationModal text='test body' show={true}  handleClose={handleClose} />
      </CartContextProvider>
    );

    const close = await findByTestId("handle-close");
    fireEvent.click(close);

    expect(handleClose).toHaveBeenCalled();
  });

  it("renders correctly", () => {
    const { container } = render(
      <CartContextProvider value={{cartItems:[]}}>
      <ConfirmationModal show={true} />
    </CartContextProvider>
      );;
    expect(container).toMatchSnapshot();
  });
});
