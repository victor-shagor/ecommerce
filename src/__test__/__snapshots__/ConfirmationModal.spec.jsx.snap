// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component renders correctly 1`] = `<div />`;
